src/detect-engine.c

```diff
@@ -987,12 +987,37 @@ int DetectBufferTypeMaxId(void)
     return g_buffer_type_id;
 }
 
+static void DetectBufferAddTransformData(DetectBufferType *map)
+{
+    for (int i = 0; i < map->transforms.cnt; i++) {
+        const TransformData *t = &map->transforms.transforms[i];
+        if (sigmatch_table[t->transform].TransformId) {
+            sigmatch_table[t->transform].TransformId(
+                    &map->xform_id[i].id_data, &map->xform_id[i].id_data_len, t->options);
+            SCLogDebug("transform identity data: [%p] \"%s\" [%d]", map->xform_id[i].id_data,
+                    (char *)map->xform_id[i].id_data, map->xform_id[i].id_data_len);
+        }
+    }
+}
+
 static uint32_t DetectBufferTypeHashNameFunc(HashListTable *ht, void *data, uint16_t datalen)
 {
     const DetectBufferType *map = (DetectBufferType *)data;
     uint32_t hash = hashlittle_safe(map->name, strlen(map->name), 0);
-    hash += hashlittle_safe((uint8_t *)&map->transforms, sizeof(map->transforms), 0);
+
+    // Add the transform data
+    // - Collect transform id and position
+    // - Collect identity data, if any
+    for (int i = 0; i < map->transforms.cnt; i++) {
+        const TransformData *t = &map->transforms.transforms[i];
+        int tval = t->transform + i;
```

Comment:
Why are we adding i ?

If I have 
1. transform id `42`, with its id_data being `43, 44`
2. transform id `42`, with its id_data being `43` and then transform id `43` + i ==1

Both will get same hash even if they are different

To make it less ambiguous, the hash could be
1. u8 number of transforms
2. all transform ids
3. all transform identity datas, prefixed by their u32 length

(that is kind of what you are doing in `DetectBufferTypeCompareNameFunc `)

Reply:
Good observation.

The "intent" is for `tval` to indicate it'itss position in the (potential) transform array. Thus, <transform-value> + <positional-value>

Reply:
> The "intent" is for tval to indicate it'itss position in the (potential) transform array. Thus, +

We do not need it, right ?
`DetectBufferTypeCompareNameFunc` does not do that, but it still uses the order of the transforms

Reply:
I don't think it's harmful and expresses the intent.

Yes -- the comparison function doesn't need to do that as it is already considers the transform order.

Reply:
> I don't think it's harmful and expresses the intent.

It does not express the intent and seems misleading to me

Reply:
I'll update the hash function to add additional values to the hash.

Reply:
The important point is that the hash should not serialize different transforms to the same serialized buffer